compiler/rustc_typeck/src/check/cast.rs

```diff
@@ -884,12 +935,16 @@ impl<'a, 'tcx> CastCheck<'tcx> {
     fn check_addr_ptr_cast(
         &self,
         fcx: &FnCtxt<'a, 'tcx>,
+        i: ty::cast::IntTy,
         m_cast: TypeAndMut<'tcx>,
     ) -> Result<CastKind, CastError> {
         // ptr-addr cast. pointer must be thin.
-        match fcx.pointer_kind(m_cast.ty, self.span)? {
-            None => Err(CastError::UnknownCastPtrKind),
-            Some(PointerKind::Thin) => Ok(CastKind::AddrPtrCast),
+        match (i, fcx.pointer_kind(m_cast.ty, self.span)?) {
+            (_, None) => Err(CastError::UnknownCastPtrKind),
+            (_, Some(PointerKind::Thin)) => Ok(CastKind::AddrPtrCast),
+            (ty::cast::IntTy::U(ty::UintTy::Usize), Some(ptr)) if ptr.maybe_wide() => {
```

Comment:
```suggestion
            (IntTy::U(ty::UintTy::Usize | ty::UintTy::Isize), Some(ptr)) if ptr.maybe_wide() => {
```
Same diagnostics are relevant to `isize` too.

Reply:
I will see how to make this work with `ty::cast::IntTy::I`.

Reply:
The code is updated to create the `UsizeToFatCast` for isize casts, but not to say "isize" in error messages (also the internal naming and comments are not updated).